lib/Sema/CSApply.cpp

```diff
@@ -627,6 +627,7 @@ namespace {
       if (isa<TypeDecl>(decl) && !isa<ModuleDecl>(decl)) {
         auto typeExpr = TypeExpr::createImplicitHack(
             loc.getBaseNameLoc(), adjustedFullType->getMetatypeInstanceType(), ctx);
+        typeExpr->setImplicit(implicit);
```

Comment:
Could you please elaborate on why this is necessary?

Reply:
In this case even though the type is written directly, this codepath creates it implicit in the AST, even though (AFAIUI) it is definitely explicit as it is written in the file (unlike other implicit cases with type inference etc). The thing that I'm wary about with this is I don't know, besides the index, what other implications this has

Reply:
I understand what you mean regarding the type being provided by the user but if we have to synthesize something we should mark it as `implicit`, having a valid loc here should be all that matters.

Reply:
this is probably just me not knowing about how this works entirely, but what qualifies for us synthesizing something? because i guess in this case the only LOC difference being an `import` to hit the implicit vs explicit case of this makes it seem to me like we're not synthesizing anything major here. Also implicit vs explicit is the only difference in the AST, I guess I assume more would be different if major things are synthesized?

Reply:
do you think this is on the right track or should i look for another solution here?

Reply:
IMO this seems reasonable enough to me, if the type was written explicitly by the user, the AST node representing it should be explicit. That being said, I think a better way of doing this would be to use `TypeExpr::createForDecl` if `implicit` is `false`, which will create a proper `SimpleIdentTypeRepr` for it. This nicely matches the logic in `resolveDeclRefExpr`:

https://github.com/apple/swift/blob/0f201443acf3b12975b63a49611dea8131c007df/lib/Sema/PreCheckExpr.cpp#L666-L676

Arguably the implicit case should also use `TypeExpr::createImplicitForDecl` too.

Reply:
Done! I didn't touch the other branch thought since I'm not sure the full impact of that